plugins/Installation/HostPortExtractor.php

```diff
@@ -0,0 +1,182 @@
+<?php
+
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link    https://matomo.org
+ * @license https://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Plugins\Installation;
+
+use InvalidArgumentException;
+
+class HostPortExtractor
+{
+    public $host;
+    public $port;
+
+    private function __construct(string $host, string $port)
+    {
+        $this->host = $host;
+        $this->port = $port;
+    }
+
+    /**
+     * Extracts the correctly formatted host and port values from the user-provided
+     * database server host.
+     *
+     * @param string $dbHost The user provided host to extract values from
+     * @return HostPortExtractor|null Extracted Host and Port, null if nothing can be extracted
+     */
+    public static function extract(string $dbHost): ?HostPortExtractor
+    {
+        try {
+            if (self::isIPv6($dbHost)) {
+                return self::extractIPv6($dbHost);
+            } elseif (self::isUnixSocket($dbHost)) {
+                return self::extractUnixSocket($dbHost);
+            } elseif (self::isIPWithPort($dbHost)) {
+                return self::extractIPAndPort($dbHost);
+            } else {
+                return null;
+            }
+        } catch (InvalidArgumentException $e) {
+            return null;
+        }
+    }
+
+    /**
+     * Determines if the provided host is correctly formatted IPv6
+     *
+     * @param string $dbHost The user provided database server host
+     * @return boolean Whether the provided host is correct IPv6
+     */
+    private static function isIPv6(string $dbHost): bool
+    {
+        // IPv6 addresses must contain '[' & ']' to be considered valid
+        if (strpos($dbHost, '[') !== false && strpos($dbHost, ']') !== false) {
```

Comment:
Perhaps we can just directly use regex to check the string starts and ends with `[` and `]` and extract the text at the same time?

With the current implementation this string would be considered valid `abc [::] def`.

Reply:
You can add strings like that into the test and then adjust the implementation until it passes.

Reply:
I have removed the first check for square brackets and improved the following regex to do the necessary checks. I have added a test for this scenario.